signer.go

```diff
@@ -2,18 +2,93 @@ package cose
 
 import (
 	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/rsa"
+	"fmt"
 	"io"
 )
 
+// Signer is an interface for private keys to sign COSE signatures.
 type Signer interface {
+	// Algorithm returns the signing algorithm associated with the private key.
 	Algorithm() Algorithm
+
+	// Sign signs digest with the private key, possibly using entropy from rand.
+	// The resulting signature should follow RFC 8152 section 8.
+	//
+	// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-8
 	Sign(rand io.Reader, digest []byte) ([]byte, error)
 }
 
+// NewSigner returns a signer with a given signing key.
+// The signing key can be a golang built-in crypto private key, a key in HSM, or
+// a remote KMS.
+//
+// Developers are encouraged to implement the `cose.Signer` interface instead of
+// the `crypto.Signer` interface for better performance.
+//
+// All signing keys implementing `crypto.Signer`` with `Public()` outputing a
+// public key of type `*rsa.PublicKey`, `*ecdsa.PublicKey`, or
+// `ed25519.PublicKey` are accepted.
+//
+// Note: `*rsa.PrivateKey`, `*ecdsa.PrivateKey`, and `ed25519.PrivateKey`
+// implement `crypto.Signer`.
 func NewSigner(alg Algorithm, key crypto.Signer) (Signer, error) {
-	panic("not implemented")
+	switch alg {
+	case AlgorithmES256, AlgorithmES384, AlgorithmES512:
+		vk, ok := key.Public().(*ecdsa.PublicKey)
+		if !ok {
+			return nil, fmt.Errorf("%v: %w", alg, ErrAlgorithmMismatch)
+		}
+		if sk, ok := key.(*ecdsa.PrivateKey); ok {
+			return &ecdsaKeySigner{
+				alg: alg,
+				key: sk,
+			}, nil
+		}
+		return &ecdsaCryptoSigner{
+			alg:    alg,
+			key:    vk,
+			signer: key,
+		}, nil
+	default:
+		return nil, ErrAlgorithmNotSupported
+	}
 }
 
-func NewSignerWithEphemeralKey(alg *Algorithm) (Signer, crypto.PrivateKey, error) {
-	panic("not implemented")
+// NewSignerWithEphemeralKey returns a Signer with a generated ephemeral key.
+// The main purpose of this function is for demo and testing.
+// Use `NewSigner` instead if a special key generation method is required.
+func NewSignerWithEphemeralKey(alg Algorithm) (Signer, crypto.PrivateKey, error) {

```

Comment:
Passing `rand io.Reader` from outside rather than using the global `rand.Reader` would ease getting deterministic keys.

Reply:
No magic "deterministic keys" by `randutil.MaybeReadByte()`. Details see https://github.com/golang/go/issues/38548.

Reply:
Not for production code, but for testing purposes it is common to use a "zero source", where all the generated random numbers are just 0, and this case is not affected by `randutil.MaybeReadByte()` non-determinism.

Reply:
`NewSignerWithEphemeralKey` was dropped.